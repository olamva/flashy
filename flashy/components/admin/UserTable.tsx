import { ActionIcon, Anchor, Avatar, Badge, ComboboxItem, Group, Loader, Select, Table, Text, rem } from "@mantine/core";
import { IconPencil, IconTrash } from "@tabler/icons-react";

import { User } from "@/app/types/user";
import { setUpdateUserRoles } from "@/app/utils/firebase";
import { notifications } from "@mantine/notifications";
import { useSession } from "next-auth/react";
import { AppRouterInstance } from "next/dist/shared/lib/app-router-context.shared-runtime";
import { useRouter } from "next/navigation";
import { useState } from "react";
import { ConfirmDeleteUser } from "../user/ConfirmationModal";

const jobColors: Record<string, string> = {
  admin: "cyan",
  user: "green",
};

type UserTableProps = {
  users: User[];
};

const norwegianRoleNames: Record<string, string> = {
  admin: "Admin",
  user: "Bruker",
};

const deleteUserFromCollection = async (actionUser: User | undefined, deleteUser: User, router: AppRouterInstance) => {
  ConfirmDeleteUser({ user: deleteUser, expires: null }, actionUser == deleteUser, router);
};

export function UsersTable({ users }: UserTableProps) {
  const { data: session, update } = useSession();
  const router = useRouter();
  const [editingUserId, setEditingUserId] = useState<string | null>(null);
  const [userRoles, setUserRoles] = useState<Record<string, ComboboxItem>>(() => {
    const roles: Record<string, ComboboxItem> = {};
    users.forEach((user) => {
      roles[user.id] = { value: user.role, label: norwegianRoleNames[user.role] };
    });
    return roles;
  });

  const roleOptions = [
    { value: "user", label: "Bruker" },
    { value: "admin", label: "Admin" },
  ];

  if (!session) {
    return <Loader color="blue" size={48} />;
  }

  const actionUser = session.user;

  const handleRoleChange = (user: User, newRole: ComboboxItem | null) => {
    if (!newRole) return;

    setUpdateUserRoles(actionUser, user.id, newRole)
      .then(() => {
        setUserRoles((prevRoles) => ({ ...prevRoles, [user.id]: newRole }));
        setEditingUserId(null);

        notifications.show({
          title: "Rolle endret",
          message: `Rollen til ${user.name} er nÃ¥ ${norwegianRoleNames[newRole.label]}`,
          color: "green",
        });

        if (session.user.id === user.id && newRole.value !== "admin") {
          update({
            ...session,
            user: { ...session.user, role: newRole.value },
          });
          router.push("/");
        }
      })
      .catch((error) => {
        notifications.show({
          title: "Noe gikk galt",
          message: error.message,
          color: "red",
        });
      });
  };

  const rows = users.map((user) => (
    <Table.Tr key={user.id}>
      <Table.Td>
        <Group gap="sm">
          <Avatar size={30} src={user.image} radius={30} />
          <Text fz="sm" fw={500}>
            {user.name}
          </Text>
        </Group>
      </Table.Td>
      <Table.Td w={200}>
        {editingUserId === user.email ? (
          <Select data={roleOptions} value={userRoles[user.id].value} onChange={(value) => handleRoleChange(user, value ? { value, label: norwegianRoleNames[value] } : null)} />
        ) : (
          <Badge color={jobColors[userRoles[user.id]?.value || user.role]} variant="light">
            {userRoles[user.id]?.label || user.role}
          </Badge>
        )}
      </Table.Td>
      <Table.Td>
        <Anchor component="button" size="sm">
          {user.email}
        </Anchor>
      </Table.Td>
      <Table.Td>
        <Group gap={0} justify="flex-end">
          <ActionIcon variant="subtle" color="gray" onClick={() => setEditingUserId(user.email)}>
            <IconPencil style={{ width: rem(16), height: rem(16) }} stroke={1.5} />
          </ActionIcon>
          <ActionIcon variant="subtle" color="red" onClick={() => deleteUserFromCollection(actionUser, user, router)}>
            <IconTrash style={{ width: rem(16), height: rem(16) }} stroke={1.5} />
          </ActionIcon>
        </Group>
      </Table.Td>
    </Table.Tr>
  ));

  return (
    <Table.ScrollContainer minWidth="100%">
      <Table verticalSpacing="sm">
        <Table.Thead>
          <Table.Tr>
            <Table.Th>Navn</Table.Th>
            <Table.Th>Rolle</Table.Th>
            <Table.Th>Email</Table.Th>
            <Table.Th />
          </Table.Tr>
        </Table.Thead>
        <Table.Tbody>{rows}</Table.Tbody>
      </Table>
    </Table.ScrollContainer>
  );
}
