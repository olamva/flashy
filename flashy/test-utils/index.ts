export * from "@testing-library/react";
export * from "./mock";
export { render } from "./render";
