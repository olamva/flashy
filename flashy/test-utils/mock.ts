// Overrides/Mocks router to prevent errors in tests
jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      prefetch: () => null,
    };
  },
  usePathname() {
    return "/";
  },
}));

// Mocks useSession hook
// jest.mock("next-auth/react", () => ({
//     useSession: jest.fn(() => ({ data: { user: dummyAdmin } }))
// }));

export {};
