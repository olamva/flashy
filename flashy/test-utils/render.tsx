import { MantineProvider } from '@mantine/core';
import { render as testingLibraryRender } from '@testing-library/react';

// From official Mantine documentation
export function render(ui: React.ReactNode) {
    return testingLibraryRender(<>{ui}</>, {
        wrapper: ({ children }: { children: React.ReactNode }) => (
            <MantineProvider>{children}</MantineProvider>
        ),
    });
}