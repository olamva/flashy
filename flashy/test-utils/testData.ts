import { FlashcardSet, FlashcardView } from "@/app/types/flashcard"
import { User } from "next-auth"

export const dummyFlashcard: FlashcardSet = {
  id: "1",
  title: "dummy",
  numViews: 2,
  numOfLikes: 2,
  numOfFavorites: 4,
  numOfComments: 3,
  createdAt: new Date(),
  popularityScore: 2,
}

export const dummyUser: User = {
  id: 'abc',
  email: "test@test.com",
  name: "Ola Nordmann",
  image: "",
  role: "user",
  age: 20,
  username: "olanordmann",
}

export const dummyAdmin: User = {
  id: 'abcd',
  email: "test@test.com",
  name: "Ola Nordmann Admin",
  image: "",
  role: "admin",
  age: 20,
  username: "olanordmannadmin",
}

export const dummyViews: FlashcardView[] = [
  {
    id: "1",
    front: "Front1",
    back: "Back1",
  },
  {
    id: "2",
    front: "Front2",
    back: "Back2",
  },
  {
    id: "3",
    front: "Front3",
    back: "Back3",
  },
]