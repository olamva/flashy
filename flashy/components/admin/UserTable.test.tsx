import { render } from '@/test-utils';
import { dummyAdmin, dummyUser } from '@/test-utils/testData';
import { useSession } from "next-auth/react";
import { UsersTable } from './UserTable';

jest.mock("next-auth/react", () => ({
    useSession: jest.fn()
}));

describe("UserTable Snapshot Test", () => {
    it("Check if snapshot matches", () => {
        (useSession as jest.Mock).mockImplementation(() => ({ data: { user: dummyAdmin } }));

        const { asFragment } = render(<UsersTable users={[dummyAdmin, dummyUser]} />);
        expect(asFragment()).toMatchSnapshot();
    });
});