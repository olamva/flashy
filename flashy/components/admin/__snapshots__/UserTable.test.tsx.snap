// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTable Snapshot Test Check if snapshot matches 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    class="m-a100c15 mantine-TableScrollContainer-scrollContainer m-d57069b5 mantine-ScrollArea-root"
    style="--table-min-width: 100%; --sa-corner-width: 0px; --sa-corner-height: 0px;"
  >
    <div
      class="m-c0783ff9 mantine-ScrollArea-viewport"
      data-offset-scrollbars="x"
      data-scrollbars="xy"
      style="overflow-x: scroll; overflow-y: scroll;"
    >
      <div
        style="min-width: 100%; display: table;"
      >
        <div
          class="m-62259741 mantine-TableScrollContainer-scrollContainerInner"
        >
          <table
            class="m-b23fa0ef mantine-Table-table"
            style="--table-vertical-spacing: var(--mantine-spacing-sm);"
          >
            <thead
              class="m-b242d975 mantine-Table-thead"
            >
              <tr
                class="m-4e7aa4fd mantine-Table-tr"
                data-with-row-border="true"
              >
                <th
                  class="m-4e7aa4f3 mantine-Table-th"
                >
                  Navn
                </th>
                <th
                  class="m-4e7aa4f3 mantine-Table-th"
                >
                  Rolle
                </th>
                <th
                  class="m-4e7aa4f3 mantine-Table-th"
                >
                  Email
                </th>
                <th
                  class="m-4e7aa4f3 mantine-Table-th"
                />
              </tr>
            </thead>
            <tbody
              class="m-b2404537 mantine-Table-tbody"
            >
              <tr
                class="m-4e7aa4fd mantine-Table-tr"
                data-with-row-border="true"
              >
                <td
                  class="m-4e7aa4ef mantine-Table-td"
                >
                  <div
                    class="m-4081bf90 mantine-Group-root"
                    style="--group-gap: var(--mantine-spacing-sm); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;"
                  >
                    <div
                      class="m-f85678b6 mantine-Avatar-root"
                      style="--avatar-size: calc(1.875rem * var(--mantine-scale)); --avatar-radius: calc(1.875rem * var(--mantine-scale));"
                    >
                      <span
                        class="m-104cd71f mantine-Avatar-placeholder"
                      >
                        <svg
                          data-avatar-placeholder-icon="true"
                          fill="none"
                          viewBox="0 0 15 15"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            clip-rule="evenodd"
                            d="M0.877014 7.49988C0.877014 3.84219 3.84216 0.877045 7.49985 0.877045C11.1575 0.877045 14.1227 3.84219 14.1227 7.49988C14.1227 11.1575 11.1575 14.1227 7.49985 14.1227C3.84216 14.1227 0.877014 11.1575 0.877014 7.49988ZM7.49985 1.82704C4.36683 1.82704 1.82701 4.36686 1.82701 7.49988C1.82701 8.97196 2.38774 10.3131 3.30727 11.3213C4.19074 9.94119 5.73818 9.02499 7.50023 9.02499C9.26206 9.02499 10.8093 9.94097 11.6929 11.3208C12.6121 10.3127 13.1727 8.97172 13.1727 7.49988C13.1727 4.36686 10.6328 1.82704 7.49985 1.82704ZM10.9818 11.9787C10.2839 10.7795 8.9857 9.97499 7.50023 9.97499C6.01458 9.97499 4.71624 10.7797 4.01845 11.9791C4.97952 12.7272 6.18765 13.1727 7.49985 13.1727C8.81227 13.1727 10.0206 12.727 10.9818 11.9787ZM5.14999 6.50487C5.14999 5.207 6.20212 4.15487 7.49999 4.15487C8.79786 4.15487 9.84999 5.207 9.84999 6.50487C9.84999 7.80274 8.79786 8.85487 7.49999 8.85487C6.20212 8.85487 5.14999 7.80274 5.14999 6.50487ZM7.49999 5.10487C6.72679 5.10487 6.09999 5.73167 6.09999 6.50487C6.09999 7.27807 6.72679 7.90487 7.49999 7.90487C8.27319 7.90487 8.89999 7.27807 8.89999 6.50487C8.89999 5.73167 8.27319 5.10487 7.49999 5.10487Z"
                            fill="currentColor"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </span>
                    </div>
                    <p
                      class="mantine-focus-auto m-b6d8b162 mantine-Text-root"
                      style="font-weight: 500;"
                    >
                      Ola Nordmann Admin
                    </p>
                  </div>
                </td>
                <td
                  class="m-4e7aa4ef mantine-Table-td"
                >
                  <div
                    class="m-347db0ec mantine-Badge-root"
                    data-variant="light"
                    style="--badge-bg: var(--mantine-color-cyan-light); --badge-color: var(--mantine-color-cyan-light-color); --badge-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
                  >
                    <span
                      class="m-5add502a mantine-Badge-label"
                    >
                      Admin
                    </span>
                  </div>
                </td>
                <td
                  class="m-4e7aa4ef mantine-Table-td"
                >
                  <button
                    class="mantine-focus-auto m-849cf0da m-b6d8b162 mantine-Text-root mantine-Anchor-root"
                    data-size="sm"
                    data-underline="hover"
                    style="--text-fz: var(--mantine-font-size-sm); --text-lh: var(--mantine-line-height-sm);"
                  >
                    test@test.com
                  </button>
                </td>
                <td
                  class="m-4e7aa4ef mantine-Table-td"
                >
                  <div
                    class="m-4081bf90 mantine-Group-root"
                    style="--group-gap: 0rem; --group-align: center; --group-justify: flex-end; --group-wrap: wrap;"
                  >
                    <button
                      class="mantine-focus-auto mantine-active m-8d3f4000 mantine-ActionIcon-root m-87cf2631 mantine-UnstyledButton-root"
                      data-variant="subtle"
                      style="--ai-bg: transparent; --ai-hover: var(--mantine-color-gray-light-hover); --ai-color: var(--mantine-color-gray-light-color); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
                      type="button"
                    >
                      <span
                        aria-hidden="true"
                        class="m-302b9fb1 mantine-ActionIcon-loader"
                      >
                        <span
                          class="m-b34414df m-5ae2e3c mantine-Loader-root"
                          style="--loader-size: calc(var(--ai-size) * 0.55); --loader-color: var(--ai-color);"
                        />
                      </span>
                      <span
                        class="m-8d3afb97 mantine-ActionIcon-icon"
                      >
                        <svg
                          class="tabler-icon tabler-icon-pencil"
                          fill="none"
                          height="24"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="1.5"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4"
                          />
                          <path
                            d="M13.5 6.5l4 4"
                          />
                        </svg>
                      </span>
                    </button>
                    <button
                      class="mantine-focus-auto mantine-active m-8d3f4000 mantine-ActionIcon-root m-87cf2631 mantine-UnstyledButton-root"
                      data-variant="subtle"
                      style="--ai-bg: transparent; --ai-hover: var(--mantine-color-red-light-hover); --ai-color: var(--mantine-color-red-light-color); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
                      type="button"
                    >
                      <span
                        aria-hidden="true"
                        class="m-302b9fb1 mantine-ActionIcon-loader"
                      >
                        <span
                          class="m-b34414df m-5ae2e3c mantine-Loader-root"
                          style="--loader-size: calc(var(--ai-size) * 0.55); --loader-color: var(--ai-color);"
                        />
                      </span>
                      <span
                        class="m-8d3afb97 mantine-ActionIcon-icon"
                      >
                        <svg
                          class="tabler-icon tabler-icon-trash"
                          fill="none"
                          height="24"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="1.5"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M4 7l16 0"
                          />
                          <path
                            d="M10 11l0 6"
                          />
                          <path
                            d="M14 11l0 6"
                          />
                          <path
                            d="M5 7l1 12a2 2 0 0 0 2 2h8a2 2 0 0 0 2 -2l1 -12"
                          />
                          <path
                            d="M9 7v-3a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v3"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                </td>
              </tr>
              <tr
                class="m-4e7aa4fd mantine-Table-tr"
                data-with-row-border="true"
              >
                <td
                  class="m-4e7aa4ef mantine-Table-td"
                >
                  <div
                    class="m-4081bf90 mantine-Group-root"
                    style="--group-gap: var(--mantine-spacing-sm); --group-align: center; --group-justify: flex-start; --group-wrap: wrap;"
                  >
                    <div
                      class="m-f85678b6 mantine-Avatar-root"
                      style="--avatar-size: calc(1.875rem * var(--mantine-scale)); --avatar-radius: calc(1.875rem * var(--mantine-scale));"
                    >
                      <span
                        class="m-104cd71f mantine-Avatar-placeholder"
                      >
                        <svg
                          data-avatar-placeholder-icon="true"
                          fill="none"
                          viewBox="0 0 15 15"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            clip-rule="evenodd"
                            d="M0.877014 7.49988C0.877014 3.84219 3.84216 0.877045 7.49985 0.877045C11.1575 0.877045 14.1227 3.84219 14.1227 7.49988C14.1227 11.1575 11.1575 14.1227 7.49985 14.1227C3.84216 14.1227 0.877014 11.1575 0.877014 7.49988ZM7.49985 1.82704C4.36683 1.82704 1.82701 4.36686 1.82701 7.49988C1.82701 8.97196 2.38774 10.3131 3.30727 11.3213C4.19074 9.94119 5.73818 9.02499 7.50023 9.02499C9.26206 9.02499 10.8093 9.94097 11.6929 11.3208C12.6121 10.3127 13.1727 8.97172 13.1727 7.49988C13.1727 4.36686 10.6328 1.82704 7.49985 1.82704ZM10.9818 11.9787C10.2839 10.7795 8.9857 9.97499 7.50023 9.97499C6.01458 9.97499 4.71624 10.7797 4.01845 11.9791C4.97952 12.7272 6.18765 13.1727 7.49985 13.1727C8.81227 13.1727 10.0206 12.727 10.9818 11.9787ZM5.14999 6.50487C5.14999 5.207 6.20212 4.15487 7.49999 4.15487C8.79786 4.15487 9.84999 5.207 9.84999 6.50487C9.84999 7.80274 8.79786 8.85487 7.49999 8.85487C6.20212 8.85487 5.14999 7.80274 5.14999 6.50487ZM7.49999 5.10487C6.72679 5.10487 6.09999 5.73167 6.09999 6.50487C6.09999 7.27807 6.72679 7.90487 7.49999 7.90487C8.27319 7.90487 8.89999 7.27807 8.89999 6.50487C8.89999 5.73167 8.27319 5.10487 7.49999 5.10487Z"
                            fill="currentColor"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </span>
                    </div>
                    <p
                      class="mantine-focus-auto m-b6d8b162 mantine-Text-root"
                      style="font-weight: 500;"
                    >
                      Ola Nordmann
                    </p>
                  </div>
                </td>
                <td
                  class="m-4e7aa4ef mantine-Table-td"
                >
                  <div
                    class="m-347db0ec mantine-Badge-root"
                    data-variant="light"
                    style="--badge-bg: var(--mantine-color-green-light); --badge-color: var(--mantine-color-green-light-color); --badge-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
                  >
                    <span
                      class="m-5add502a mantine-Badge-label"
                    >
                      Bruker
                    </span>
                  </div>
                </td>
                <td
                  class="m-4e7aa4ef mantine-Table-td"
                >
                  <button
                    class="mantine-focus-auto m-849cf0da m-b6d8b162 mantine-Text-root mantine-Anchor-root"
                    data-size="sm"
                    data-underline="hover"
                    style="--text-fz: var(--mantine-font-size-sm); --text-lh: var(--mantine-line-height-sm);"
                  >
                    test@test.com
                  </button>
                </td>
                <td
                  class="m-4e7aa4ef mantine-Table-td"
                >
                  <div
                    class="m-4081bf90 mantine-Group-root"
                    style="--group-gap: 0rem; --group-align: center; --group-justify: flex-end; --group-wrap: wrap;"
                  >
                    <button
                      class="mantine-focus-auto mantine-active m-8d3f4000 mantine-ActionIcon-root m-87cf2631 mantine-UnstyledButton-root"
                      data-variant="subtle"
                      style="--ai-bg: transparent; --ai-hover: var(--mantine-color-gray-light-hover); --ai-color: var(--mantine-color-gray-light-color); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
                      type="button"
                    >
                      <span
                        aria-hidden="true"
                        class="m-302b9fb1 mantine-ActionIcon-loader"
                      >
                        <span
                          class="m-b34414df m-5ae2e3c mantine-Loader-root"
                          style="--loader-size: calc(var(--ai-size) * 0.55); --loader-color: var(--ai-color);"
                        />
                      </span>
                      <span
                        class="m-8d3afb97 mantine-ActionIcon-icon"
                      >
                        <svg
                          class="tabler-icon tabler-icon-pencil"
                          fill="none"
                          height="24"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="1.5"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4"
                          />
                          <path
                            d="M13.5 6.5l4 4"
                          />
                        </svg>
                      </span>
                    </button>
                    <button
                      class="mantine-focus-auto mantine-active m-8d3f4000 mantine-ActionIcon-root m-87cf2631 mantine-UnstyledButton-root"
                      data-variant="subtle"
                      style="--ai-bg: transparent; --ai-hover: var(--mantine-color-red-light-hover); --ai-color: var(--mantine-color-red-light-color); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
                      type="button"
                    >
                      <span
                        aria-hidden="true"
                        class="m-302b9fb1 mantine-ActionIcon-loader"
                      >
                        <span
                          class="m-b34414df m-5ae2e3c mantine-Loader-root"
                          style="--loader-size: calc(var(--ai-size) * 0.55); --loader-color: var(--ai-color);"
                        />
                      </span>
                      <span
                        class="m-8d3afb97 mantine-ActionIcon-icon"
                      >
                        <svg
                          class="tabler-icon tabler-icon-trash"
                          fill="none"
                          height="24"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="1.5"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M4 7l16 0"
                          />
                          <path
                            d="M10 11l0 6"
                          />
                          <path
                            d="M14 11l0 6"
                          />
                          <path
                            d="M5 7l1 12a2 2 0 0 0 2 2h8a2 2 0 0 0 2 -2l1 -12"
                          />
                          <path
                            d="M9 7v-3a1 1 0 0 1 1 -1h4a1 1 0 0 1 1 1v3"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
